### PR TITLE
Fixes nrc/rustfmt#187

### PR DESCRIPTION
**Pull request by myromeu**

### CODE REVIEW

**Comment by marcusklaas** [comment]
Thanks for the PR! There's a couple more instances where `.len() == 0` is used. Could you replace those as well?

**Comment by myromeu** [comment]
I don't understand 662 line..
What should I do with this?

**Comment by nrc** [comment]
@myromeu `lifetimes.is_empty() && tys.is_empty()`

**Comment by myromeu** [comment]
'+' similarly '&&' ?

**Comment by nrc** [comment]
No, not quite.

So, in this case there are two lists and we want to check that both are empty. Currently we do this by taking the length of each list and adding them together and checking to see if they are zero. This relies on the fact that if a = 0 and b = 0 then a+b = 0. The `&&` is logical and (i.e., `a && b` is true if both `a` and `b` are true). `+` is normal addition of two numbers.

**Comment by myromeu** [comment]
nrc, thank you. I didn't think of immediately (:

**Comment by myromeu** [comment]
Its completed.

**Comment by nrc** [comment]
Thanks, looks good! It needs a rebase, could you try to fix the Git history too please? This could all be one commit and shouldn't include the default.toml commit (this is probably due to merging rather rebasing your local branch. I imagine the easiest way to fix this is to start with a new branch and cherry-pick your two commits from the existing branch, then squash those two commits, then rebase to upstream/master). If you could give the commit a meaningful description too, and just reference #187 in the body of the commit message, that would be awesome.

**Comment by myromeu** [comment]
I apologize for the long absence. I understand what you want. I'll try to do it..

**Comment by myromeu** [comment]
I can't fix that. 
I added my commits in new branch. After I merged changes (resolve all changes). And ran 'cargo test'..
That ended with output:
'''
   Compiling rustfmt v0.0.1 (file:///home/hasher/projects/rust/rustfmt)
     Running target/debug/rustfmt-70b05d62e82ba6b1

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

     Running target/debug/rustfmt-f6206b8e3f3f6f09

running 7 tests
test issues::find_issue ... ok
test comment::test_find_uncommented ... ok
test comment::comment_end ... ok
test comment::format_comments ... ok
test issues::find_unnumbered_issue ... ok
test issues::issue_type ... ok
test utils::power_rounding ... ok

test result: ok. 7 passed; 0 failed; 0 ignored; 0 measured

     Running target/debug/system-8aee14f10fe2178e

running 2 tests
test system_tests ... ok
test idempotence_tests ... FAILED

failures:

---- idempotence_tests stdout ----
        Testing 'tests/target/issue-64.rs'...

Testing 'tests/target/attrib.rs'...

Testing 'tests/target/match.rs'...

Testing 'tests/target/loop.rs'...

Testing 'tests/target/mod-1.rs'...

Testing 'tests/target/imports-reorder.rs'...
Testing 'tests/target/string-lit.rs'...

Testing 'tests/target/struct_lits.rs'...

Testing 'tests/target/skip.rs'...

Testing 'tests/target/tuple.rs'...

Testing 'tests/target/imports.rs'...
Rustfmt failed at tests/target/imports.rs:5: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/imports.rs:6: line exceeded maximum length (sorry)

Testing 'tests/target/doc.rs'...

Testing 'tests/target/closure.rs'...
WARNING: tests/target/closure.rs:29: found TODO 

Testing 'tests/target/expr.rs'...
Rustfmt failed at tests/target/expr.rs:12: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/expr.rs:60: line exceeded maximum length (sorry)

Testing 'tests/target/attrib-extern-crate.rs'...

Testing 'tests/target/fn.rs'...

Testing 'tests/target/hello.rs'...

Testing 'tests/target/enum.rs'...

Testing 'tests/target/structs.rs'...

Testing 'tests/target/comments-fn.rs'...

Testing 'tests/target/trait.rs'...

Testing 'tests/target/multiple.rs'...
Rustfmt failed at tests/target/multiple.rs:25: line exceeded maximum length (sorry)

Testing 'tests/target/struct_lits_visual.rs'...
Rustfmt failed at tests/target/struct_lits_visual.rs:20: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:21: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:22: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:23: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:24: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:25: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:26: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:28: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:29: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:30: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:31: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:32: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:33: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/struct_lits_visual.rs:34: line exceeded maximum length (sorry)

Testing 'tests/target/long-fn-1.rs'...

Testing 'tests/target/paths.rs'...
Rustfmt failed at tests/target/paths.rs:5: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/paths.rs:7: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/paths.rs:8: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/paths.rs:9: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/paths.rs:10: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/paths.rs:13: line exceeded maximum length (sorry)
Rustfmt failed at tests/target/paths.rs:17: line exceeded maximum length (sorry)

Testing 'tests/target/fn-simple.rs'...

Testing 'tests/target/expr-visual-indent.rs'...

Testing 'tests/target/mod-2.rs'...

Testing 'tests/target/other.rs'...

Testing 'tests/system.rs'...
WARNING: tests/system.rs:163: found TODO 
WARNING: tests/system.rs:177: found TODO 

Testing 'src/bin/rustfmt.rs'...

Testing 'src/lib.rs'...
thread 'idempotence_tests' panicked at 'arithmetic operation overflowed', src/expr.rs:1011

thread 'idempotence_tests' panicked at 'Downcast failed.', ../src/libcore/option.rs:331



failures:
    idempotence_tests

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured
'''

**Comment by cassiersg** [comment]
It's difficult to say anything without the code. Can you push the buggy merge on this PR ?

**Comment by myromeu** [comment]
Yeah
Code was pushed.

**Comment by cassiersg** [comment]
There is a lot of conflict marks remaining here, the code doesn't compile. Please push the code with conflicts resolved (the code that caused the error you shown above).

**Comment by myromeu** [comment]
It builds now, but not completes tests...

**Comment by cassiersg** [comment]
This solves the bug (this was an unsigned integer computation underflowing).
```diff
--- a/src/expr.rs
+++ b/src/expr.rs
@@ -1008,7 +1008,7 @@ fn rewrite_unary_op(context: &RewriteContext,
         ast::UnOp::UnNeg => "-",
     };
 
-    let subexpr = try_opt!(expr.rewrite(context, width - operator_str.len(), offset));
+    let subexpr = try_opt!(expr.rewrite(context, try_opt!(width.checked_sub(operator_str.len())), offset));
 
     Some(format!("{}{}", operator_str, subexpr))
 }
```

**Comment by myromeu** [comment]
I solved it.

File: src/expr.rs [hunk 1]
```diff
@@ -1008,7 +1008,8 @@ fn rewrite_unary_op(context: &RewriteContext,
         ast::UnOp::UnNeg => "-",
     };
 
-    let subexpr = try_opt!(expr.rewrite(context, width - operator_str.len(), offset));
+    let subexpr = try_opt!(expr.rewrite(context,
+    try_opt!(width.checked_sub(operator_str.len())), offset));
```
Comment (cassiersg) [inline src/expr.rs]: nit: indent

**Comment by myromeu** [comment]
In fact, this change does not apply to action)

**Comment by cassiersg** [comment]
Thanks again for this PR !

**Pull request merged**